lib/AST/ASTContext.cpp

```diff
@@ -11182,3 +11182,46 @@ APFixedPoint ASTContext::getFixedPointMin(QualType Ty) const {
   assert(Ty->isFixedPointType());
   return APFixedPoint::getMin(getFixedPointSemantics(Ty));
 }
+
+RecordDecl *ASTContext::getCachedTypeApp(const RecordDecl *Base, ArrayRef<const Type *> TypeArgs) {
+  auto it = CachedTypeApps.find(std::make_pair(Base, TypeArgs));
+  if (it == CachedTypeApps.end()) return nullptr;
+  return it->second;
+}
+
+void ASTContext::addCachedTypeApp(const RecordDecl *Base, ArrayRef<const Type *> TypeArgs, RecordDecl *Inst) {
```

Comment:
Can RecordDecl *Inst be marked as const?

Reply:
Is it safe to assume "Base" and "Inst" are never going to be nullptrs? Perhaps, we should either dynamically check for null (defense in depth) or add asserts.

Similar observation for methods below in the PR.

Reply:
> Can RecordDecl *Inst be marked as const?
Maybe, but it'd require quite a few changes: the const leaks out to the users and their callers and their callers, etc. I'll leave it be for now.

> Is it safe to assume "Base" and "Inst" are never going to be nullptrs? Perhaps, we should either dynamically check for null (defense in depth) or add asserts.
> 
> Similar observation for methods below in the PR.

Added assertions.